gpu-aware-scheduling/pkg/gpuscheduler/resource_map.go

```diff
@@ -0,0 +1,145 @@
+package gpuscheduler
+
+import (
+	"errors"
+
+	"k8s.io/klog/v2"
+)
+
+const (
+	minAllowedInput = 0
+)
+
+// Errors.
+var (
+	errOverflow = errors.New("integer overflow")
+	errInput    = errors.New("input error")
+)
+
+// Map of resources. name -> resource amount.
+type resourceMap map[string]int64
+
+func (rm resourceMap) newCopy() resourceMap {
+	mapCopy := make(resourceMap, len(rm))
+
+	mapCopy.copyFrom(rm)
+
+	return mapCopy
+}
+
+func (rm resourceMap) copyFrom(src resourceMap) {
+	for key := range src {
+		rm[key] = src[key]
+	}
+}
+
+// addRM adds the resources of the src resourceMap and returns nil.
+// If there is an error adding the resources, nothing will be added.
+func (rm resourceMap) addRM(src resourceMap) error {
+	mapCopy := rm.newCopy()
+	// check with the copy that the src fits
+	for key, value := range src {
+		err := mapCopy.add(key, value)
+		if err != nil {
+			klog.Error("addRM failed")
+
+			return err
+		}
+	}
+
+	rm.copyFrom(mapCopy)
+
+	return nil
+}
+
+// subtractRM removes the resources of the src resourceMap and returns nil.
+// If there is an error removing the resources, nothing will be removed.
+// If any resource amount would go negative, it is set to zero.
+func (rm resourceMap) subtractRM(src resourceMap) error {
+	mapCopy := rm.newCopy()
+	// check with the copy that the src can be subtracted
+	for key, value := range src {
+		err := mapCopy.subtract(key, value)
+		if err != nil {
+			klog.Error("subtractRM failed")
+
+			return err
+		}
+	}
+
+	rm.copyFrom(mapCopy)
+
+	return nil
+}
+
+// add adds a resource to the map.
+// It assumes being used on a sane resource map with positive values.
+func (rm resourceMap) add(key string, value int64) error {
+	if value < minAllowedInput {
+		klog.Error("bad input for add, key:", key)
+
+		return errInput
+	}
+
```

Comment:
This code has a lot of vertical whitespace - it makes it harder to read, at least for me :)

Reply:
cilint with all the bells an whistles here